background.js

```diff
@@ -0,0 +1,471 @@
+	if (listeners[tabId]) {
+		chrome.webRequest.onBeforeRequest.removeListener(listeners[tabId]);
+		delete listeners[tabId];
+		queryTeamscale('stop', tabId);
+	}
+});
+
+function registerInitialListeners(tabId) {
+	fetchStoredConfiguration();
+	cacheTeamscaleSessionCookie();
+
+	initialListeners[tabId * 2] = details => resolveAzureDevOpsSessionIdListener(details, tabId);
+	initialListeners[tabId * 2 + 1] = details => onStartListener(details, tabId);
+
+	const listenerFilterOptions = {
+		urls: currentUriFilter,
+		types: ["xmlhttprequest"],
+		tabId: tabId
+	};
+
+	chrome.webRequest.onBeforeSendHeaders.addListener(initialListeners[tabId * 2], listenerFilterOptions, ["requestHeaders", "extraHeaders"]);
+	chrome.webRequest.onBeforeRequest.addListener(initialListeners[tabId * 2 + 1], listenerFilterOptions, ["requestBody"]);
+}
+
+function resolveAzureDevOpsSessionIdListener(details, tabId) {
+	if (!isTestRunnerApiCall(details)) {
+		return;
+	}
+
+	const headers = details.requestHeaders;
+
+	for (let i = 0; i < headers.length; i++) {
+		if (headers[i].name.toLowerCase() !== 'X-TFS-Session'.toLowerCase()) {
+			continue;
+		}
+
+		tabToSessionMap[tabId] = headers[i].value;
+	}
+
+	// self delete this listener
+	chrome.webRequest.onBeforeSendHeaders.removeListener(initialListeners[tabId * 2]);
+	if (!tabToSessionMap[tabId]) {
+		throw 'Could not obtain Azure DevOps Session ID.';
+	}
+}
+
+function onStartListener(details, tabId) {
+	if (!isTestRunnerApiCall(details)) {
+		return;
+	}
+
+	let workItemId;
+	let apiCallUserInfoUri;
+	if (details.url.endsWith(API_ON_PREMISE_CALL_OPEN_TEST_RUNNER)) {
+		workItemId = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(details.requestBody.raw[0].bytes))).workItemIds;
+		apiCallUserInfoUri = details.url.substring(0, details.url.length - API_ON_PREMISE_CALL_OPEN_TEST_RUNNER.length) +
+			'/_api/_common/GetUserProfile?__v=5';
+	} else {
+		workItemId = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(details.requestBody.raw[0].bytes))).context.properties.workItemIds;
+		apiCallUserInfoUri = details.url.substring(0, details.url.length - API_SERVICES_CALL_OPEN_TEST_RUNNER.length) +
+			'/_api/_common/GetUserProfile?__v=5';
+	}
+
+	tabToTestCaseMap[tabId] = workItemId;
+
+	resolveUserNameOfTesterAndTriggerRecordingStart(apiCallUserInfoUri, tabId);
+
+	// self delete this listener
+	chrome.webRequest.onBeforeRequest.removeListener(initialListeners[tabId * 2 + 1]);
+}
+
+/**
+ * This listener is attached to web requests in tabs that are assumed to belong to an Azure DevOps test runner and
+ * parses needed information from update calls (regarding a test run) to Azure DevOps.
+ */
+function testRunUpdateCallListener(details, tabId, tab) {
+	if (!details.url.endsWith(API_CALL_UPDATE_TEST_RUN_SUFFIX)) {
+		return;
+	}
+
+	let updateRequest = JSON.parse(JSON.parse(String.fromCharCode.apply(
+		null, new Uint8Array(details.requestBody.raw[0].bytes))).updateRequests)[0];
```

Comment:
What if we cannot parse it as JSON and then try to get the first element? Should we also check this case?